collect_app/src/test/java/org/odk/collect/android/location/client/LocationClientProviderTest.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2020 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.odk.collect.android.location.client;
+
+import android.content.Context;
+import android.util.Log;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.odk.collect.android.utilities.PlayServicesChecker;
+import org.robolectric.RobolectricTestRunner;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+@RunWith(RobolectricTestRunner.class)
+public class LocationClientProviderTest {
+    @Rule
+    public MockitoRule rule = MockitoJUnit.rule();
+
+    @Mock
+    public Context context;
+
+    @Mock
+    public PlayServicesChecker playServicesChecker;
+
+    @Test
+    public void fusedLocationClient_returnedWhenPlayServicesAvailable() {
+        when(playServicesChecker.isGooglePlayServicesAvailable(any())).thenReturn(true);
+
+        // Could import Robolectric Shadows Play Services to actually build a GoogleFusedLocationClient
```

Comment:
I agree about not pulling in Play Services shadows. One potentially simple way to make this test a little nicer would be to pass a `GoogleAPIClientLocationProvider` interface to `LocationClientProvider` in someway so it could be mocked/faked and not actually call the play services code. That would also let us write tests for `GoogleFusedLocationProvider` without any extra Robolectric stuff. 

So many interfaces though! It's frustrating not having any way to express lambdas in method signatures. #3886 adds a poly fill for Java 8's [Supplier](https://docs.oracle.com/javase/8/docs/api/java/util/function/Supplier.html)  (we could be using it as a general "provider" interface) which could make this all a bit simpler. You could think about introducing that here?

Reply:
Hopefully 87439ff3 is roughly what you had in mind. There are so many levels of provider client client providers here it's feeling a little silly but I agree the test looks much better.